### PR TITLE
Creating python classes for HTML code generation

### PR DESCRIPTION
**Pull request by shravan1k**
### Administrative Info
Monday Board ID: 940408300
Make sure your branch name conforms to: `<feature/staging/hotfix/...>/[username]/[Monday Item ID]-[3-4 word description separated by dashes]`. Otherwise, please rename your branch and create a new PR.

### Changes
What changes did you make?
- Created python classes to represent variable aspects of the html code
- Tested out to see if those objects were being properly accessed when generating html.

### Testing
How did you confirm your changes worked? 
- Generated new html file and opened in browser.

The nolinebody.py file is not complete yet, I used it for testing the objects and if they were being properly accessed before generating the html.

### Confirmation of Change 
Upload a screenshot, if possible. Otherwise, please provide instructions on how to see the change.

| Testing using python objects |
| --- |
| Screen Shot 2021-01-24 at 6 35 24 PM |

### CODE REVIEW

File: newsletter.html [hunk 1]
```diff
@@ -0,0 +1,48 @@
+<!DOCTYPE html>
```
Comment (wes-chen) [inline newsletter.html:1]: Let's remove this file from the pull request, as it isn't relevant to what we are doing, maybe even add it into the `.gitignore`?

File: py/bodynoline.py [hunk 2]
```diff
@@ -15,5 +15,20 @@ def _bodynoline():
     Raises:
         TODO
     """
-
+    with table(cls="sk-table"):
```
Comment (wes-chen) [inline py/bodynoline.py:18]: if we're not ready to commit this file yet, let's go ahead and leave changes on `bodynoline.py` out of the pull request.

File: py/entry_objects.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
+class UpcomingEvent:
+    """
+    Represents one entry of an upcoming event
```
Comment (wes-chen) [inline py/entry_objects.py:3]: In this description, can we get a better description of what exactly an "upcoming event" entails?

File: py/entry_objects.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+    3. body (array of line or text objects) - Each element in the array represents a new line
+    """
+    def __init__(self, title, details, body):
+        self.title = title
+        self.details = details
+        self.body = body
+    
+class Entry:
+    """
+    Represents one entry of all other sections other than upcoming events
```
Comment (wes-chen) [inline py/entry_objects.py:16]: Same as above, this description really does not give me much information here.

File: py/entry_objects.py [hunk 5]
```diff
@@ -0,0 +1,75 @@
+class UpcomingEvent:
```
Comment (wes-chen) [inline py/entry_objects.py:1]: if we have Milestones, Exec Updates, and AlumniAdvice as children of Entries, why is an UpcomingEvent not a child?

File: py/entry_objects.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    def __init__(self, title, body):
+        super().__init__(title, body)
+        
+class Text:
+    """
+    Represents a portion of text of an entry's body.
+    Attributes:
+    1. text (string) - The actual text that needs to be displayed
+    2. linebreaks (integer) - The number of empty lines that needs to follow the text
```
Comment (wes-chen) [inline py/entry_objects.py:51]: Having linebreaks as an attribute feels odd to me, wouldn't people just add "\n\n\n\n\n..." at the end of their strings to add more space? Even so, I feel like this is still bad practice to just slap linebreaks to do spacing when we should be customizing css - HTML is meant to determine structure of the contents of a document, while CSS determines how the contents are arranged. Having linebreaks appended at the end breaks this principle.

File: py/entry_objects.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+    """
+    def __init__(self, text, url, linebreaks):
+        super().__init__(text, linebreaks)
+        self.url = url
+
+class Image:
+    """
+    Represents an image
+    Attributes:
+    1. url (string) - The url of the image to be displayed.
```
Comment (wes-chen) [inline py/entry_objects.py:72]: What about `alt`? from w3schools: alt - Specifies an alternate text for the image, if the image for some reason cannot be displayed

File: py/newsletter.py [hunk 8]
```diff
@@ -19,14 +20,21 @@
 
 # create the document body
 with doc:
+    alumni_desc = ["Harvard Business School MBA Candidate", "Formerly Salesforce Software Engineer", "UCSD Class of 2017 - Co-Founder of TSE"]
```
Comment (wes-chen) [inline py/newsletter.py:23]: again, if we're just playing around, let's remove these changes from the PR.

**Review by wes-chen — CHANGES_REQUESTED** [review]
Some initial comments, great first try

File: py/bodynoline.py [hunk 9]
```diff
@@ -1,7 +1,7 @@
 import dominate
 from dominate.tags import *
 
-def _bodynoline():
+def _bodynoline(title, entries):
     """Returns a body section with no vertical lines. TODO: edit this description
```
Comment (wes-chen) [inline py/bodynoline.py:5]: Add a description on the params here please: :)

**Review by wes-chen — CHANGES_REQUESTED** [review]
almost there

File: py/bodynoline.py (hunk 9 above)
Comment (shravan1k) [reply py/bodynoline.py:5]: added

File: py/bodynoline.py [hunk 10]
```diff
@@ -1,19 +1,19 @@
...
 
     TODO: A description here, according to Google's Python style guide https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
 
     Args:
-        TODO
```
Comment (wes-chen) [inline py/bodynoline.py:10]: finish all the TODOs please :)

**Review by wes-chen — CHANGES_REQUESTED** [review]

File: py/bodynoline.py (hunk 10 above)
Comment (shravan1k) [reply py/bodynoline.py:10]: done

File: py/entry_objects.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
+class Entry:
+    """
+    Each section on the newsletter, such as upcoming events and milestones, can have multiple entries.
+    This class is used to represent a single entry which consists of a title and a description.
+    Attributes:
+    1. title (string) - The title of the entry.
+    2. body (list of Line, Text or Image objects) - Each element in the list represents a new line.
```
Comment (wes-chen) [inline py/entry_objects.py:7]: does this refer to the `LineBreak` object or the `Link` object? (possible misspell)

File: py/newsletter.py [hunk 12]
```diff
@@ -24,13 +25,19 @@
...
 with doc:
     # enclose all of the elements within div elements (currently set to ef-styling)
-    with div(cls='wrapper').add(div(cls='container')).add(table(cls='table outer')):
-        _head()
-        _greetings(greet_title)
-        _bodylineleft()
-        _bodylineright()
-        _bodynoline()
-        _foot()
+    with div(cls='sk-wrapper').add(div(cls='sk-webkit')).add(table(cls='sk-table sk-outer sk-height')):
```
Comment (wes-chen) [inline py/newsletter.py:28]: are the `sk-*` classnames still relevant? or can we switch to all being on one css file with out the prefixes

**Review by wes-chen — CHANGES_REQUESTED** [review]
a few more changes

File: py/entry_objects.py (hunk 11 above)
Comment (shravan1k) [reply py/entry_objects.py:7]: Yeah i meant Link objects, but will add Linebreak as well.

File: py/newsletter.py (hunk 12 above)
Comment (shravan1k) [reply py/newsletter.py:28]: Yeah i can make those changes

Comment (shravan1k) [reply py/newsletter.py]: I removed the sk's for the classes in the main py file for now. I'll gradually remove the sk's for the other classes as I use them in the different python files for each section. This is so I can gradually test and make sure changing every class name doesn't affect the display.

**Review by wes-chen — APPROVED** [review]
great work!

File: py/bodynoline.py [hunk 13]
```diff
@@ -1,19 +1,20 @@
...
-
-    TODO: A description here, according to Google's Python style guide https://github.com/google/styleguide/blob/gh-pages/pyguide.md#38-comments-and-docstrings
+def _bodynoline(title, entries):
+    """Returns the HTML for a centered body section with no vertical lines. The section has a 
+    title, such as Alumni Advice or Member Spotlight, as well as a body section that contains
+    text along with images.
 
     Args:
-        TODO
+        title (string) - The title of the newsletter section.
```
Comment (eliasfang) [inline py/bodynoline.py:10]: Maybe add an example of a title and entries that are passed in

File: py/newsletter.py [hunk 14]
```diff
@@ -24,13 +25,19 @@
...
 with doc:
     # enclose all of the elements within div elements (currently set to ef-styling)
-    with div(cls='wrapper').add(div(cls='container')).add(table(cls='table outer')):
-        _head()
-        _greetings(greet_title)
-        _bodylineleft()
-        _bodylineright()
-        _bodynoline()
-        _foot()
+    with div(cls='wrapper').add(div(cls='container')).add(table(cls='outer height')):
```
Comment (eliasfang) [inline py/newsletter.py:28]: How come you needed to change this to how it is now?

**Review by eliasfang — CHANGES_REQUESTED** [review]
Sorry to add more to do but I just had some clarification edits. I tried looking for things that were wrong per Wesley's request but nothing looks wrong to me.

File: py/newsletter.py (hunk 14 above)
Comment (shravan1k) [reply py/newsletter.py:28]: Before, all the sections were under the table tag but there were no rows or cells being created.

File: py/bodynoline.py (hunk 13 above)
Comment (shravan1k) [reply py/bodynoline.py:10]: Good idea!

File: py/newsletter.py (hunk 14 above)
Comment (eliasfang) [reply py/newsletter.py:28]: I see. If you look at my head and greetings I actually create the rows and cells in there but I'll take your changes into consideration.

Comment (shravan1k) [reply py/newsletter.py:28]: Oh okay, sounds good. The way I thought about it was that it's probably better to create the rows and cells with the table itself.

**Review by eliasfang — APPROVED** [review]
Much clearer!

**Pull request merged**